### PR TITLE
Add bpb and n_bytes to metric logging

### PR DESCRIPTION
**Pull request by EntilZha**
Summary:

Test Plan:

### CODE REVIEW

File: bytelatent/metrics.py [hunk 1]
```diff
@@ -54,14 +55,24 @@ class LoggingArgs(BaseModel):
...
+        self.fs = fs
         self.args = args
 
     def open(self):
         if self.jsonl_writer is None:
-            self.jsonl_writer = open(self.outdir, "a")
+            if self.fs is None:
+                self.jsonl_writer = open(self.outdir, "a")
+            else:
+                self.jsonl_writer = self.fs.open(self.outdir, "a")
```
Comment (sriniiyer) [inline bytelatent/metrics.py:75]: What functionality does this add?

File: bytelatent/train.py [hunk 2]
```diff
@@ -588,20 +626,33 @@ def train(args: TrainArgs):
...
-                    f"  loss: {round(loss.item(),4):>7}"
+                    f"  loss: step={round(loss.item(),4):>7} avg={avg_loss}"
+                    f"  bpb: {avg_bpb:3f}"
                     f"  grad: {grad_norm:.2e}"
                     f"  flops: {FLOPS:.2e}"
                     f"  wps: {wps:.2e}"
                     f"  iter: {curr_iter_time:>7}"
                     f"  data: {data_load_time:>5}"
                     f"  lr: {curr_lr:.2e}"
+                    f"  n_bytes={total_n_bytes}"
```
Comment (sriniiyer) [inline bytelatent/train.py:647]: this is confusing, we need a way to indicate per_gpu vs. across all gpus

File: bytelatent/train.py [hunk 3]
```diff
@@ -588,20 +626,33 @@ def train(args: TrainArgs):
...
+                )
+                total_n_bytes = dist_sum(n_bytes, reduce_dtype=torch.bfloat16)
+                avg_bpb = total_tok_loss / math.log(2) / total_n_bytes
+                avg_loss = dist_mean(np.mean(step_losses).item())
                 logger.info(
                     f"step: {train_state.step}"
                     f"  acc: {train_state.acc_step}"
-                    f"  loss: {round(loss.item(),4):>7}"
+                    f"  loss: step={round(loss.item(),4):>7} avg={avg_loss}"
+                    f"  bpb: {avg_bpb:3f}"
```
Comment (sriniiyer) [inline bytelatent/train.py:640]: This is inconsistent, we normally report bpb per gpu

File: bytelatent/metrics.py (hunk 1 above)
Comment (EntilZha) [reply bytelatent/metrics.py]: Eventually (once we fix `torch.distributed.checkpointing.save` to work with `fsspec`), all of `dump_dir` should be compatible with writing to nfs or s3/blob store.

File: bytelatent/train.py [hunk 4]
```diff
@@ -568,50 +594,105 @@ def train(args: TrainArgs):
...
+                # Below semantics are:
+                # step=Metrics at a step
+                # interval=Metrics averaged across the logging interval
+                # local=On one rank
+                # global=Across all ranks
                 logger.info(
                     f"step: {train_state.step}"
                     f"  acc: {train_state.acc_step}"
-                    f"  loss: {round(loss.item(),4):>7}"
+                    f"  loss: [step_local={round(step_loss_per_gpu, 4):>7} interval_local={round(interval_loss_per_gpu, 4):>7} step_global={round(step_loss_across_gpus, 4):>7} interval_global={round(interval_loss_across_gpus, 4):>7}]"
```
Comment (sriniiyer) [inline bytelatent/train.py:676]: this is too many things - I think interval local and interval global is good enough
